### PR TITLE
Add zstd (un)archiver support

### PR DESCRIPTION
**Pull request by pleeplop**
Hi everyone,

This pull request provides support for zstd format.
https://github.com/facebook/zstd

(De)compression is provided by zstd-jni library (BSD license) and apache commons-compress.
https://github.com/luben/zstd-jni

Let me know if you have any feedback.

### CODE REVIEW

File: src/main/java/org/codehaus/plexus/archiver/zstd/PlexusIoZstdResourceCollection.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    protected InputStream getInputStream( File file ) throws IOException
+    {
+        FileInputStream fileIs = new FileInputStream( file );
+        try
+        {
+            final InputStream result = ZstdUnArchiver.getZstdInputStream( fileIs );
+            fileIs = null;
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply src/main/java/org/codehaus/plexus/archiver/zstd/PlexusIoZstdResourceCollection.java]: It is fixed

File: src/main/java/org/codehaus/plexus/archiver/zstd/ZstdUnArchiver.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import javax.annotation.Nonnull;
+import javax.inject.Named;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static org.codehaus.plexus.archiver.util.Streams.*;
+
+/**
+ * @author pleeplop
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply src/main/java/org/codehaus/plexus/archiver/zstd/ZstdUnArchiver.java]: It is fixed

File: src/main/java/org/codehaus/plexus/archiver/zstd/ZstdArchiver.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    {
+    }
+
+    /**
+     * Set compression level
+     */
+    public void setLevel( Integer level )
+        throws ArchiverException
+    {
+        compressor.setLevel(level);
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply src/main/java/org/codehaus/plexus/archiver/zstd/ZstdArchiver.java]: I also fixed missing spaces in other files.

File: src/main/java/org/codehaus/plexus/archiver/zstd/ZstdCompressor.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    private Integer level;
+
+    private ZstdCompressorOutputStream zstdOut;
+
+    public ZstdCompressor()
+    {
+    }
+
+    public void setLevel(Integer level) {
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply src/main/java/org/codehaus/plexus/archiver/zstd/ZstdCompressor.java]: It's fixed

File: src/main/java/org/codehaus/plexus/archiver/zstd/ZstdUnArchiver.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import org.codehaus.plexus.archiver.AbstractUnArchiver;
+import org.codehaus.plexus.archiver.ArchiverException;
+
+import javax.annotation.Nonnull;
+import javax.inject.Named;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static org.codehaus.plexus.archiver.util.Streams.*;
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply src/main/java/org/codehaus/plexus/archiver/zstd/ZstdUnArchiver.java]: I removed wildcard imports.

File: pom.xml [hunk 6]
```diff
@@ -141,6 +141,11 @@
       <version>3.23.1</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.github.luben</groupId>
+      <artifactId>zstd-jni</artifactId>
```
[the remark this answers is not in the document]
Comment (pleeplop) [reply pom.xml]: That's right, it should be runtime scope.
I moved the dependency after org.tukaani.xz

**Comment by plamentotev** [comment]
Thanks for the contribution

**Pull request merged**